corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+
+----
+
+Elastic Client Adapters
+-----------------------
+
+The ``corehq.apps.es.client`` module encapsulates the CommCare HQ Elasticsearch
+client adapters. It implements a high-level Elasticsearch client protocol
+necessary to accomplish all interactions with the backend Elasticsearch cluster.
+Client adapters are split into two usage patterns, the "Management Adapter" and
+"Document Adapters".
+
+.. toctree::
+
+    Management Adapter
+    Document Adapters
+    Code Documentation
+
+
+Management Adapter
+''''''''''''''''''
+
+There is only one management adapter, ``ElasticManageAdapter``. This adapter is
+used for performing all cluster management tasks such as creating and updating
+indices and their mappings, changing index settings, changing cluster settings,
+etc.  This functionality is split into a separate class for a few reasons:
+
+1. The management adapter is responsible for low-level Elastic operations which
+   document adapters should never be performing because the scope of a document
+   adapter does not extend beyond a single index.
+2. Elasticsearch 5+ implements security features which limit the kinds of
+   operations a connection can be used for. The separation in these client
+   adapter classes is designed to fit into that model.
+
+The management adapter does not need any special parameters to work with, and
+can be instantiated and used directly:
+
+.. code-block:: python
+
+    adapter = ElasticManageAdapter()
+    adapter.index_create("books")
+    mapping = {"properties": {
+        "author": {"type": "text"},
+        "title": {"type": "text"},
+        "published": {"type": "date"},
+    }}
+    adapter.index_put_mapping("books", "book", mapping)
+    adapter.index_refresh("books")
+    adapter.index_delete("books")
+
+
+Document Adapters
+'''''''''''''''''
+
+Document adapters are created on a per-index basis and include specific
+properties and functionality necessary for maintaining a single type of "model"
+document in a single index.  Each index in Elasticsearch needs to have a
+cooresponding ``ElasticDocumentAdapter`` subclass which defines how the Python
+model is applied to that specific index.  At the very least, a document adapter
+must define the following:
+
+- An ``_index_name`` attribute whose value is the name of the Elastic index
+  used by the adapter. This attribute must be private to support proper index
+  naming between production code and tests.
+- A ``type`` attribute whose value is the name is the Elastic ``_type`` for
+  documents used by the adapter.
+- A ``mapping`` which defines the structure and properties for documents managed
+  by the adapter.
+- A ``from_python()`` classmethod which can convert a Python model object into the
+  JSON-serializable format for writing into the adapter's index.
+
+The combination of ``(index_name, type)`` constrains the document adapter to
+a specific HQ document mapping.  Comparing an Elastic cluster to a Postgres
```

Comment:
Yes, I'm aware of this, but in order to facilitate a smooth transition from 2.4 (which requires them) to newer a version, they're an unfortunate reality in the interim.

Reply:
Yeah, sorry I did not mean to suggest you should remove them from this PR. I was wondering if you had a sense how easy it will be to remove them later, and if there are steps we could take, either in enforcing how they are used, or just in documentation that will make it easier to make that change, since that is another key part of the upgrade starting in version 5. That might just be noting in this section for other devs that types are deprecated going forward and new functionality should be developed that rely on them, since this new documentation implies they are a key part of the ES model, which is hopefully only true for the shortest possible time, and we should not count on that for any feature.

Reply:
> I was wondering if you had a sense how easy it will be to remove them later

It will be trivial to remove them later. The adapter will soon be the _only_ code even aware of their existence, so they should be easy to strip out entirely in a very small change set.